Use unique_ptr in DofMap::merge_ghost_functor_outputs() API

jwpeterson opened the pull request:
An API change, so if necessary, we could implement a backwards-compatible shim for this. If nothing in MOOSE calls this directly I think we should probably be OK to skip that. Also, this uses a C++14 capability related to the `is_transparent` type, but @roystgnr and I tried it on a bunch of older compilers and it seems to be pretty well-supported.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1022746/) on 749c02f wanted to post the following:


## Coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c0b238</th>
      <th colspan=4>#3238 749c02</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>55.76%</td>
      <td>55.76%</td>
      <td>+0.00%</td>
      <td>42.86%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>44482</td>
      <td>44481</td>
      <td>-1</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>35290</td>
      <td>35286</td>
      <td>-4</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/libmesh/docs/PRs/3238/coverage_diff/diff.html)

[Full coverage report](https://mooseframework.inl.gov/libmesh/docs/PRs/3238/coverage)

## Warnings

- New new line coverage rate 42.86% is less than the suggested 90.0%



This comment will be updated on new commits.

jwpeterson commented:
The failure here in `mesh/custom_partitioner.group/custom_linear_partitioner_displacement` definitely seems real and related to my changes, but I'm not exactly sure what the issue would be.
```
../src/base/dof_map.C, line 1587, compiled Apr 20 2022 at 10:58:462
Assertion `ghost_coupling->size() == n_var' failed.
ghost_coupling->size() = 2
n_var = 1
```

roystgnr commented on include/utils/utility.h:
```diff
@@ -464,6 +465,48 @@ T ReverseBytes::operator() (T & data) const
+ * Struct which defines a custom comparison object that
+ * can be used with std::sets of std::unique_ptrs
+ */
+template <class T>
```
Do we need this to be templated around T?  Or would templating the member functions work?

jwpeterson replied:
I don't know actually... I just assumed this was how you were supposed to do it, but I can check...

jwpeterson replied:
OK, I tried templating just the member functions and it still worked. This makes the `typedef` declaration nicer since you don't have to repeat the type twice... so I'll go ahead and make that change.

roystgnr commented:
It looks to me like a find() is returning end() (and thus prompting us to create a duplicate of an existing functor) rather than actually finding what it should, but I don't see how.

Maybe we ought to have a couple unit tests for the new comparator?  This seems like the sort of thing we really want to catch at the start of "make check", not deep into ghosting functor operations.

jwpeterson commented:
> Maybe we ought to have a couple unit tests for the new comparator?

OK. I also (force-pushed) a change since the last failure which might fix the problem. I was trying to do fewer `const_casts` but I think I ended up introducing a bug... not sure but the CI will let us know.

jwpeterson commented:
OK, nice, there was a real bug found by that MOOSE test but it is now fixed. 

I can add some unit tests for this. I have the test code that I used to develop everything initially, plus I expanded upon that to verif that stuff like the following works as expected (although not something that can be used safely in real code):
```
// Test that inserting the same underlying pointer (after moving) fails
std::set<std::unique_ptr<int>, CompareUnderlying> s3;
auto four = std::make_unique<int>(4);
s3.insert(std::move(four)); // steals resources of four
s3.insert(std::move(four)); // inserts nullptr underlying value, moved-from unique_ptr is guaranteed to be nullptr 
s3.insert(std::move(four)); // verify that insert does not happen, nullptr underlying value already exists
```

roystgnr commented:
I'd feel safer with unit tests but this is still a big improvement as-is.
